File: codigo/src/Quadtree.cpp
#include "../include/Quadtree.h"

Quadtree::Quadtree(AABB bounds, int lvl) {
    boundary = bounds;
    level = lvl;
    divided = false;
    for (int i = 0; i < 4; i++) children[i] = nullptr;
}

Quadtree::~Quadtree() {
    clear();
}

void Quadtree::clear() {
    objects.clear();
    if (divided) {
        for (int i = 0; i < 4; i++) {
            if (children[i]) {
                delete children[i];
                children[i] = nullptr;
            }
        }
        divided = false;
    }
}

void Quadtree::subdivide() {
    float x = boundary.x;
    float y = boundary.y;
    float hw = boundary.halfW;
    float hh = boundary.halfH;
    float qHW = hw / 2.0f;
    float qHH = hh / 2.0f;

    children[0] = new Quadtree(AABB(x - qHW, y - qHH, qHW, qHH), level + 1); // NW
    children[1] = new Quadtree(AABB(x + qHW, y - qHH, qHW, qHH), level + 1); // NE
    children[2] = new Quadtree(AABB(x - qHW, y + qHH, qHW, qHH), level + 1); // SW
    children[3] = new Quadtree(AABB(x + qHW, y + qHH, qHW, qHH), level + 1); // SE

    divided = true;
}

int Quadtree::getIndex(const AABB& itemBounds) {
    int index = -1;
    float verticalMidpoint = boundary.x;
    float horizontalMidpoint = boundary.y;

    bool topQuadrant = (itemBounds.y + itemBounds.halfH < horizontalMidpoint);
    bool bottomQuadrant = (itemBounds.y - itemBounds.halfH > horizontalMidpoint);
    bool leftQuadrant = (itemBounds.x + itemBounds.halfW < verticalMidpoint);
    bool rightQuadrant = (itemBounds.x - itemBounds.halfW > verticalMidpoint);

    if (leftQuadrant) {
        if (topQuadrant) index = 0;
        else if (bottomQuadrant) index = 2;
    }
    else if (rightQuadrant) {
        if (topQuadrant) index = 1;
        else if (bottomQuadrant) index = 3;
    }

    return index;
}

bool Quadtree::insert(Entity* entity) {
    if (!boundary.intersects(entity->bounds)) return false;

    if (divided) {
        int index = getIndex(entity->bounds);
        if (index != -1) return children[index]->insert(entity);
    }

    objects.push_back(entity);

    if (objects.size() > CAPACITY && level < MAX_DEPTH) {
        if (!divided) subdivide();

        SimpleList<Entity*> movedObjects;

        objects.remove_if([&](Entity* obj) {
            int idx = getIndex(obj->bounds);
            if (idx != -1) {
                children[idx]->insert(obj);
                return true;
            }
            return false;
        });
    }
    return true;
}

void Quadtree::query(AABB range, SimpleList<Entity*>& found) {
    if (!boundary.intersects(range)) return;

    for (auto& obj : objects) {
        if (obj->active && range.intersects(obj->bounds)) {
            found.push_back(obj);
        }
    }

    if (divided) {
        for (int i = 0; i < 4; i++) {
            children[i]->query(range, found);
        }
    }
}

void Quadtree::debugRender(sf::RenderWindow& window) {
    sf::RectangleShape rect;

    rect.setSize(sf::Vector2f(boundary.halfW * 2.0f, boundary.halfH * 2.0f));

    rect.setPosition(boundary.x - boundary.halfW, boundary.y - boundary.halfH);

    rect.setFillColor(sf::Color::Transparent);

    sf::Uint8 colorIntensity = std::min(255, 50 + (level * 25));
    rect.setOutlineColor(sf::Color(0, colorIntensity, 255, 150));
    rect.setOutlineThickness(1.0f / (level + 1));

    window.draw(rect);

    if (divided) {
        for (int i = 0; i < 4; i++) {
            if (children[i]) {
                children[i]->debugRender(window);
            }
        }
    }
}

File: web/quadtree.js
// ============================================================
//  quadtree.js – Puerto de Quadtree.cpp y DataStructures.h
// ============================================================

class AABB {
    constructor(x, y, halfW, halfH) {
        this.x = x;
        this.y = y;
        this.halfW = halfW;
        this.halfH = halfH;
    }

    intersects(other) {
        if (Math.abs(this.x - other.x) > (this.halfW + other.halfW)) return false;
        if (Math.abs(this.y - other.y) > (this.halfH + other.halfH)) return false;
        return true;
    }

    contains(other) {
        return (
            other.x - other.halfW >= this.x - this.halfW &&
            other.x + other.halfW <= this.x + this.halfW &&
            other.y - other.halfH >= this.y - this.halfH &&
            other.y + other.halfH <= this.y + this.halfH
        );
    }
}

const QT_CAPACITY = 4;
const QT_MAX_DEPTH = 6;

class Quadtree {
    constructor(bounds, level = 0) {
        this.boundary = bounds;
        this.level = level;
        this.divided = false;
        this.objects = [];
        this.children = [null, null, null, null];
    }

    clear() {
        this.objects = [];
        if (this.divided) {
            for (let i = 0; i < 4; i++) {
                if (this.children[i]) {
                    this.children[i].clear();
                    this.children[i] = null;
                }
            }
            this.divided = false;
        }
    }

    subdivide() {
        const x = this.boundary.x;
        const y = this.boundary.y;
        const qHW = this.boundary.halfW / 2;
        const qHH = this.boundary.halfH / 2;

        this.children[0] = new Quadtree(new AABB(x - qHW, y - qHH, qHW, qHH), this.level + 1); // NW
        this.children[1] = new Quadtree(new AABB(x + qHW, y - qHH, qHW, qHH), this.level + 1); // NE
        this.children[2] = new Quadtree(new AABB(x - qHW, y + qHH, qHW, qHH), this.level + 1); // SW
        this.children[3] = new Quadtree(new AABB(x + qHW, y + qHH, qHW, qHH), this.level + 1); // SE
        this.divided = true;
    }

    _getIndex(itemBounds) {
        const vMid = this.boundary.x;
        const hMid = this.boundary.y;

        const top    = (itemBounds.y + itemBounds.halfH) < hMid;
        const bottom = (itemBounds.y - itemBounds.halfH) > hMid;
        const left   = (itemBounds.x + itemBounds.halfW) < vMid;
        const right  = (itemBounds.x - itemBounds.halfW) > vMid;

        if (left) {
            if (top)    return 0;
            if (bottom) return 2;
        } else if (right) {
            if (top)    return 1;
            if (bottom) return 3;
        }
        return -1;
    }

    insert(entity) {
        if (!this.boundary.intersects(entity.bounds)) return false;

        if (this.divided) {
            const idx = this._getIndex(entity.bounds);
            if (idx !== -1) return this.children[idx].insert(entity);
        }

        this.objects.push(entity);

        if (this.objects.length > QT_CAPACITY && this.level < QT_MAX_DEPTH) {
            if (!this.divided) this.subdivide();

            this.objects = this.objects.filter(obj => {
                const idx = this._getIndex(obj.bounds);
                if (idx !== -1) {
                    this.children[idx].insert(obj);
                    return false;
                }
                return true;
            });
        }
        return true;
    }

    query(range, found) {
        if (!this.boundary.intersects(range)) return;

        for (const obj of this.objects) {
            if (obj.active && range.intersects(obj.bounds)) {
                found.push(obj);
            }
        }

        if (this.divided) {
            for (let i = 0; i < 4; i++) {
                this.children[i].query(range, found);
            }
        }
    }

    debugRender(ctx) {
        const b = this.boundary;
        const intensity = Math.min(255, 50 + this.level * 25);
        ctx.strokeStyle = `rgba(0, ${intensity}, 255, 0.6)`;
        ctx.lineWidth = Math.max(0.3, 1 / (this.level + 1));
        ctx.strokeRect(b.x - b.halfW, b.y - b.halfH, b.halfW * 2, b.halfH * 2);

        if (this.divided) {
            for (let i = 0; i < 4; i++) {
                this.children[i].debugRender(ctx);
            }
        }
    }
}
